<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>megadepth: BigWig and BAM related utilities • megadepth</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="megadepth: BigWig and BAM related utilities">
<meta property="og:description" content="This package provides an R interface to Megadepth by Christopher
    Wilks available at https://github.com/ChristopherWilks/megadepth. It is
    particularly useful for computing the coverage of a set of genomic regions
    across bigWig or BAM files. With this package, you can build base-pair
    coverage matrices for regions or annotations of your choice from BigWig
    files. Megadepth was used to create the raw files provided by
    https://bioconductor.org/packages/recount3.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">megadepth</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/megadepth.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/LieberInstitute/megadepth/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="megadepth" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#megadepth" class="anchor"></a>megadepth</h1></div>
<!-- badges: start -->

<p>The goal of <code>megadepth</code> is to provide an R interface to the command line tool <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> for BigWig and BAM related utilities created by <a href="https://twitter.com/chrisnwilks">Christopher Wilks</a>. This R package enables <strong>fast</strong> processing of BigWig files on downstream packages such as <a href="https://bioconductor.org/packages/dasper">dasper</a> and <a href="https://bioconductor.org/packages/recount3">recount3</a>. The <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> software also provides utilities for processing BAM files and extracting coverage information from them.</p>
<p>Here is an illustration on how fast <code>megadepth</code> is compared to other tools for processing local and remote BigWig files.</p>
<p><a href="https://github.com/LieberInstitute/megadepth/tree/master/analysis"><img src="https://raw.githubusercontent.com/LieberInstitute/megadepth/master/analysis/md_rt_pybw_runtime.png" width="800px"></a></p>
<p>Throughout the documentation we use a capital <code>M</code> to refer to the software by Christopher Wilks and a lower case <code>m</code> to refer to this R/Bioconductor package.</p>
<div id="installation-instructions" class="section level2">
<h2 class="hasAnchor">
<a href="#installation-instructions" class="anchor"></a>Installation instructions</h2>
<p>Get the latest stable <code>R</code> release from <a href="http://cran.r-project.org/">CRAN</a>. Then install <code>megadepth</code> using from <a href="http://bioconductor.org/">Bioconductor</a> the following code:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="op">}</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"megadepth"</span><span class="op">)</span></code></pre></div>
<p>And the development version from <a href="https://github.com/">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"LieberInstitute/megadepth"</span><span class="op">)</span></code></pre></div>
</div>
<div id="example" class="section level2">
<h2 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h2>
<p>In the following example, we install <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> with <code><a href="reference/install_megadepth.html">install_megadepth()</a></code>, which downloads a binary for your OS (Linux, Windows or macOS). We can then use with an example BigWig file to compute the coverage at a set of regions.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Load the R package</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/LieberInstitute/megadepth">"megadepth"</a></span><span class="op">)</span>

<span class="co">## Install Megadepth's pre-compiled binary on your system</span>
<span class="fu"><a href="reference/install_megadepth.html">install_megadepth</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; The latest megadepth version is 1.1.0c</span>
<span class="co">#&gt; megadepth has been installed to /home/rstudio/bin</span>

<span class="co">## Next, we locate the example BigWig and annotation files</span>
<span class="va">example_bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"tests"</span>, <span class="st">"test.bam.all.bw"</span>,
    package <span class="op">=</span> <span class="st">"megadepth"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>
<span class="va">annotation_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"tests"</span>, <span class="st">"testbw2.bed"</span>,
    package <span class="op">=</span> <span class="st">"megadepth"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co">## We can then use megadepth to compute the coverage</span>
<span class="va">bw_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/get_coverage.html">get_coverage</a></span><span class="op">(</span><span class="va">example_bw</span>, op <span class="op">=</span> <span class="st">"mean"</span>, annotation <span class="op">=</span> <span class="va">annotation_file</span><span class="op">)</span>
<span class="va">bw_cov</span>
<span class="co">#&gt; GRanges object with 4 ranges and 1 metadata column:</span>
<span class="co">#&gt;         seqnames          ranges strand |     score</span>
<span class="co">#&gt;            &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span>
<span class="co">#&gt;   [1]      chr10            0-10      * |      0.00</span>
<span class="co">#&gt;   [2]      chr10 8756697-8756762      * |     15.85</span>
<span class="co">#&gt;   [3]      chr10 4359156-4359188      * |      3.00</span>
<span class="co">#&gt;   [4] GL000219.1   168500-168620      * |      1.26</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
</div>
<div id="full-set-of-utilities" class="section level2">
<h2 class="hasAnchor">
<a href="#full-set-of-utilities" class="anchor"></a>Full set of utilities</h2>
<p><a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> is very powerful and can do a lot of different things. The R/Bioconductor package provides two functions for interfacing with <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a>, <code><a href="reference/megadepth_cmd.html">megadepth_cmd()</a></code> and <code><a href="reference/megadepth_cmd.html">megadepth_shell()</a></code>. For the first one, <code><a href="reference/megadepth_cmd.html">megadepth_cmd()</a></code>, you need to know the actual command syntax you want to use and format it accordingly. If you are more comfortable with R functions, <code><a href="reference/megadepth_cmd.html">megadepth_shell()</a></code> uses <em><a href="https://CRAN.R-project.org/package=cmdfun">cmdfun</a></em> to power this interface and capture the standard output stream into R.</p>
<p>To make it easier to use, <code>megadepth</code> includes functions that simplify the number of arguments, read in the output files, and converts them into R/Bioconductor friendly objects, such as <code><a href="reference/get_coverage.html">get_coverage()</a></code> illustrated above.</p>
<p>We hope that you’ll find <code>megadepth</code> and <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> useful for your work. If you are interested in checking how <strong>fast</strong> <code>megadepth</code> is, check out the <a href="https://github.com/LieberInstitute/megadepth/tree/master/analysis"><strong>speed analysis</strong></a> comparison against other tools. Note that the size of the files used and the number of genomic regions queried will affect the speed comparisons.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## R-like interface</span>
<span class="co">## that captures the standard output into R</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="reference/megadepth_cmd.html">megadepth_shell</a></span><span class="op">(</span>help <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "megadepth 1.1.0c"                 ""                                </span>
<span class="co">#&gt; [3] "BAM and BigWig utility."          ""                                </span>
<span class="co">#&gt; [5] "Usage:"                           "  megadepth &lt;bam|bw|-&gt; [options]"</span>

<span class="co">## Command-like interface</span>
<span class="fu"><a href="reference/megadepth_cmd.html">megadepth_cmd</a></span><span class="op">(</span><span class="st">"--help"</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; megadepth 1.1.0c</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  BAM and BigWig utility.</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  Usage:</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    megadepth &lt;bam|bw|-&gt; [options]</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  Options:</span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    -h --help                Show this screen.</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --version                Show version.</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --threads                # of threads to do: BAM decompression OR compute sums over multiple BigWigs in parallel</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                              if the 2nd is intended then a TXT file listing the paths to the BigWigs to process in parallel</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                              should be passed in as the main input file instead of a single BigWig file (EXPERIMENTAL).</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --prefix                 String to use to prefix all output files.</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --no-auc-stdout          Force all AUC(s) to be written to &lt;prefix&gt;.auc.tsv rather than STDOUT</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --no-annotation-stdout   Force summarized annotation regions to be written to &lt;prefix&gt;.annotation.tsv rather than STDOUT</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --no-coverage-stdout     Force covered regions to be written to &lt;prefix&gt;.coverage.tsv rather than STDOUT</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --keep-order             Output annotation coverage in the order chromosomes appear in the BAM/BigWig file</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                             The default is to output annotation coverage in the order chromosomes appear in the annotation BED file.</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                             This is only applicable if --annotation is used for either BAM or BigWig input.</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  BigWig Input:</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  Extract regions and their counts from a BigWig outputting BED format if a BigWig file is detected as input (exclusive of the other BAM modes):</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                            Extracts all reads from the passed in BigWig and output as BED format.</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                             This will also report the AUC over the annotated regions to STDOUT.</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                             If only the name of the BigWig file is passed in with no other args, it will *only* report total AUC to STDOUT.</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --annotation &lt;bed&gt;                      Only output the regions in this BED applying the argument to --op to them.</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --op &lt;sum[default], mean, min, max&gt;     Statistic to run on the intervals provided by --annotation</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --sums-only                             Discard coordinates from output of summarized regions</span></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --bwbuffer &lt;1GB[default]&gt;               Size of buffer for reading BigWig files, critical to use a large value (~1GB) for remote BigWigs.</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                             Default setting should be fine for most uses, but raise if very slow on a remote BigWig.</span></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  BAM Input:</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  Extract basic junction information from the BAM, including co-occurrence</span></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  If only the name of the BAM file is passed in with no other args, it will *only* report total AUC to STDOUT.</span></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --fasta              Path to the reference FASTA file if a CRAM file is passed as the input file (ignored otherwise)</span></span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         If not passed, references will be downloaded using the CRAM header.</span></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --junctions          Extract co-occurring jx coordinates, strand, and anchor length, per read</span></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         writes to a TSV file &lt;prefix&gt;.jxs.tsv</span></span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --all-junctions      Extract all jx coordinates, strand, and anchor length, per read for any jx</span></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         writes to a TSV file &lt;prefix&gt;.all_jxs.tsv</span></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --longreads          Modifies certain buffer sizes to accommodate longer reads such as PB/Oxford.</span></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --filter-in          Integer bitmask, any bits of which alignments need to have to be kept (similar to samtools view -f).</span></span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --filter-out         Integer bitmask, any bits of which alignments need to have to be skipped (similar to samtools view -F).</span></span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --add-chr-prefix     Adds "chr" prefix to relevant chromosomes for BAMs w/o it, pass "human" or "mouse".</span></span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         Only works for human/mouse references (default: off).</span></span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  Non-reference summaries:</span></span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --alts                       Print differing from ref per-base coverages</span></span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                 Writes to a CSV file &lt;prefix&gt;.alts.tsv</span></span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --include-softclip           Print a record to the alts CSV for soft-clipped bases</span></span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                 Writes total counts to a separate TSV file &lt;prefix&gt;.softclip.tsv</span></span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --only-polya                 If --include-softclip, only print softclips which are mostly A's or T's</span></span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --include-n                  Print mismatch records when mismatched read base is N</span></span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --print-qual                 Print quality values for mismatched bases</span></span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --delta                      Print POS field as +/- delta from previous</span></span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --require-mdz                Quit with error unless MD:Z field exists everywhere it's</span></span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                                 expected</span></span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --head                       Print sequence names and lengths in SAM/BAM header</span></span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  Coverage and quantification:</span></span>
<span id="cb5-63"><a href="#cb5-63" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --coverage           Print per-base coverage (slow but totally worth it)</span></span>
<span id="cb5-64"><a href="#cb5-64" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --auc                Print per-base area-under-coverage, will generate it for the genome</span></span>
<span id="cb5-65"><a href="#cb5-65" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         and for the annotation if --annotation is also passed in</span></span>
<span id="cb5-66"><a href="#cb5-66" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         Defaults to STDOUT, unless other params are passed in as well, then</span></span>
<span id="cb5-67"><a href="#cb5-67" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         if writes to a TSV file &lt;prefix&gt;.auc.tsv</span></span>
<span id="cb5-68"><a href="#cb5-68" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --bigwig             Output coverage as BigWig file(s).  Writes to &lt;prefix&gt;.bw</span></span>
<span id="cb5-69"><a href="#cb5-69" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         (also &lt;prefix&gt;.unique.bw when --min-unique-qual is specified).</span></span>
<span id="cb5-70"><a href="#cb5-70" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         Requires libBigWig.</span></span>
<span id="cb5-71"><a href="#cb5-71" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --annotation &lt;BED|window_size&gt;   Path to BED file containing list of regions to sum coverage over</span></span>
<span id="cb5-72"><a href="#cb5-72" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         (tab-delimited: chrm,start,end). Or this can specify a contiguous region size in bp.</span></span>
<span id="cb5-73"><a href="#cb5-73" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --op &lt;sum[default], mean&gt;     Statistic to run on the intervals provided by --annotation</span></span>
<span id="cb5-74"><a href="#cb5-74" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --no-index           If using --annotation, skip the use of the BAM index (BAI) for pulling out regions.</span></span>
<span id="cb5-75"><a href="#cb5-75" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         Setting this can be faster if doing windows across the whole genome.</span></span>
<span id="cb5-76"><a href="#cb5-76" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         This will be turned on automatically if a window size is passed to --annotation.</span></span>
<span id="cb5-77"><a href="#cb5-77" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --min-unique-qual &lt;int&gt;</span></span>
<span id="cb5-78"><a href="#cb5-78" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         Output second bigWig consisting built only from alignments</span></span>
<span id="cb5-79"><a href="#cb5-79" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         with at least this mapping quality.  --bigwig must be specified.</span></span>
<span id="cb5-80"><a href="#cb5-80" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         Also produces second set of annotation sums based on this coverage</span></span>
<span id="cb5-81"><a href="#cb5-81" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         if --annotation is enabled</span></span>
<span id="cb5-82"><a href="#cb5-82" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --double-count       Allow overlapping ends of PE read to count twice toward</span></span>
<span id="cb5-83"><a href="#cb5-83" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         coverage</span></span>
<span id="cb5-84"><a href="#cb5-84" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --num-bases          Report total sum of bases in alignments processed (that pass filters)</span></span>
<span id="cb5-85"><a href="#cb5-85" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --gzip               Turns on gzipping of coverage output (no effect if --bigwig is passsed),</span></span>
<span id="cb5-86"><a href="#cb5-86" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         this will also enable --no-coverage-stdout.</span></span>
<span id="cb5-87"><a href="#cb5-87" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-88"><a href="#cb5-88" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  Other outputs:</span></span>
<span id="cb5-89"><a href="#cb5-89" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --read-ends          Print counts of read starts/ends, if --min-unique-qual is set</span></span>
<span id="cb5-90"><a href="#cb5-90" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         then only the alignments that pass that filter will be counted here</span></span>
<span id="cb5-91"><a href="#cb5-91" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         Writes to 2 TSV files: &lt;prefix&gt;.starts.tsv, &lt;prefix&gt;.ends.tsv</span></span>
<span id="cb5-92"><a href="#cb5-92" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --frag-dist          Print fragment length distribution across the genome</span></span>
<span id="cb5-93"><a href="#cb5-93" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;                         Writes to a TSV file &lt;prefix&gt;.frags.tsv</span></span>
<span id="cb5-94"><a href="#cb5-94" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --echo-sam           Print a SAM record for each aligned read</span></span>
<span id="cb5-95"><a href="#cb5-95" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --ends               Report end coordinate for each read (useful for debugging)</span></span>
<span id="cb5-96"><a href="#cb5-96" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;    --test-polya         Lower Poly-A filter minimums for testing (only useful for debugging/testing)</span></span>
<span id="cb5-97"><a href="#cb5-97" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  </span></span>
<span id="cb5-98"><a href="#cb5-98" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span></code></pre></div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>Citation</h2>
<p>Below is the citation output from using <code><a href="https://rdrr.io/r/utils/citation.html">citation('megadepth')</a></code> in R. Please run this yourself to check for any updates on how to cite <strong>megadepth</strong>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/print.html">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"megadepth"</span><span class="op">)</span>, bibtex <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Zhang D, Collado-Torres L (2021). _megadepth: BigWig and BAM related</span>
<span class="co">#&gt; utilities_. doi: 10.18129/B9.bioc.megadepth (URL:</span>
<span class="co">#&gt; https://doi.org/10.18129/B9.bioc.megadepth),</span>
<span class="co">#&gt; https://github.com/LieberInstitute/megadepth - R package version 1.1.6,</span>
<span class="co">#&gt; &lt;URL: http://www.bioconductor.org/packages/megadepth&gt;.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; A BibTeX entry for LaTeX users is</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   @Manual{,</span>
<span class="co">#&gt;     title = {megadepth: BigWig and BAM related utilities},</span>
<span class="co">#&gt;     author = {David Zhang and Leonardo Collado-Torres},</span>
<span class="co">#&gt;     year = {2021},</span>
<span class="co">#&gt;     url = {http://www.bioconductor.org/packages/megadepth},</span>
<span class="co">#&gt;     note = {https://github.com/LieberInstitute/megadepth - R package version 1.1.6},</span>
<span class="co">#&gt;     doi = {10.18129/B9.bioc.megadepth},</span>
<span class="co">#&gt;   }</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Wilks C, Ahmed O, Baker DN, Zhang D, Collado-Torres L, Langmead B</span>
<span class="co">#&gt; (2020). "Megadepth: efficient coverage quantification for BigWigs and</span>
<span class="co">#&gt; BAMs." _bioRxiv_. doi: 10.1101/2020.12.17.423317 (URL:</span>
<span class="co">#&gt; https://doi.org/10.1101/2020.12.17.423317), &lt;URL:</span>
<span class="co">#&gt; https://www.biorxiv.org/content/10.1101/2020.12.17.423317v1&gt;.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; A BibTeX entry for LaTeX users is</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;   @Article{,</span>
<span class="co">#&gt;     title = {Megadepth: efficient coverage quantification for BigWigs and BAMs},</span>
<span class="co">#&gt;     author = {Christopher Wilks and Omar Ahmed and Daniel N. Baker and David Zhang and Leonardo Collado-Torres and Ben Langmead},</span>
<span class="co">#&gt;     year = {2020},</span>
<span class="co">#&gt;     journal = {bioRxiv},</span>
<span class="co">#&gt;     doi = {https://doi.org/10.1101/2020.12.17.423317},</span>
<span class="co">#&gt;     url = {https://www.biorxiv.org/content/10.1101/2020.12.17.423317v1},</span>
<span class="co">#&gt;   }</span></code></pre></div>
<p>Please note that the <code>megadepth</code> was only made possible thanks to many other R and bioinformatics software authors, which are cited either in the vignettes and/or the paper(s) describing this package.</p>
</div>
<div id="code-of-conduct" class="section level2">
<h2 class="hasAnchor">
<a href="#code-of-conduct" class="anchor"></a>Code of Conduct</h2>
<p>Please note that the <code>megadepth</code> project is released with a <a href="https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html">Contributor Code of Conduct</a>. By contributing to this project, you agree to abide by its terms.</p>
</div>
<div id="development-tools" class="section level2">
<h2 class="hasAnchor">
<a href="#development-tools" class="anchor"></a>Development tools</h2>
<ul>
<li>Continuous code testing is possible thanks to <a href="https://www.tidyverse.org/blog/2020/04/usethis-1-6-0/">GitHub actions</a> through <em><a href="https://CRAN.R-project.org/package=usethis">usethis</a></em>, <em><a href="https://CRAN.R-project.org/package=remotes">remotes</a></em>, <em><a href="https://github.com/r-hub/sysreqs">sysreqs</a></em> and <em><a href="https://CRAN.R-project.org/package=rcmdcheck">rcmdcheck</a></em> customized to use <a href="https://www.bioconductor.org/help/docker/">Bioconductor’s docker containers</a> and <em><a href="https://bioconductor.org/packages/3.12/BiocCheck">BiocCheck</a></em>.</li>
<li>Code coverage assessment is possible thanks to <a href="https://codecov.io/gh">codecov</a> and <em><a href="https://CRAN.R-project.org/package=covr">covr</a></em>.</li>
<li>The <a href="http://LieberInstitute.github.io/megadepth">documentation website</a> is automatically updated thanks to <em><a href="https://CRAN.R-project.org/package=pkgdown">pkgdown</a></em>.</li>
<li>The code is styled automatically thanks to <em><a href="https://CRAN.R-project.org/package=styler">styler</a></em>.</li>
<li>The documentation is formatted thanks to <em><a href="https://CRAN.R-project.org/package=devtools">devtools</a></em> and <em><a href="https://CRAN.R-project.org/package=roxygen2">roxygen2</a></em>.</li>
</ul>
<p>For more details, check the <code>dev</code> directory.</p>
<p>This package was developed using <em><a href="https://bioconductor.org/packages/3.12/biocthis">biocthis</a></em>.</p>
</div>
<div id="recount-project" class="section level2">
<h2 class="hasAnchor">
<a href="#recount-project" class="anchor"></a><code>ReCount</code> project</h2>
<p>The <strong>main documentation website</strong> for all the <code>recount3</code>-related projects is available at <a href="https://LieberInstitute/github.io/recount3-docs"><strong>recount.bio</strong></a>. Please check that website for more information about how this R/Bioconductor package and other tools are related to each other.</p>
</div>
<div id="teams-involved" class="section level2">
<h2 class="hasAnchor">
<a href="#teams-involved" class="anchor"></a>Teams involved</h2>
<p><em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> was made possible to <a href="https://twitter.com/dyzhang32">David Zhang</a>, the author of <em><a href="https://bioconductor.org/packages/3.12/dasper">dasper</a></em>, and a member of the <a href="https://snca.atica.um.es/">Mina Ryten</a>’s lab at UCL.</p>
<p>The <code>ReCount</code> family involves the following teams:</p>
<ul>
<li>
<a href="http://www.langmead-lab.org/">Ben Langmead’s lab</a> at JHU Computer Science</li>
<li>
<a href="https://www.hansenlab.org/">Kasper Daniel Hansen’s lab</a> at JHBSPH Biostatistics Department</li>
<li>
<a href="http://lcolladotor.github.io/">Leonardo Collado-Torres</a> and <a href="http://aejaffe.com/">Andrew E. Jaffe</a> from <a href="https://www.libd.org/">LIBD</a>
</li>
<li>
<a href="http://nellore.bio/">Abhinav Nellore’s lab</a> at OHSU</li>
<li>
<a href="http://jtleek.com/">Jeff Leek’s lab</a> at JHBSPH Biostatistics Deparment</li>
<li>Data hosted by <a href="https://www.sciserver.org/">SciServer from IDIES at JHU</a>
</li>
</ul>
<table class="table"><tbody><tr class="odd">
<td><a href="http://www.langmead-lab.org/"><img src="http://www.langmead-lab.org/wp-content/uploads/2014/01/Screen-Shot-2014-02-02-at-5.20.13-PM-1024x199.png" width="250px"></a></td>
<td><a href="https://www.libd.org/"><img src="http://lcolladotor.github.io/img/LIBD_logo.jpg" width="250px"></a></td>
<td><a href="http://nellore.bio/"><img src="https://seekvectorlogo.net/wp-content/uploads/2018/08/oregon-health-science-university-ohsu-vector-logo.png" width="250px"></a></td>
<td><a href="https://www.sciserver.org/"><img src="https://skyserver.sdss.org/dr14/en/images/sciserver_logo_inverted_vertical.png" width="250px"></a></td>
</tr></tbody></table>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Download from BIOC at <br><a href="https://www.bioconductor.org/packages/megadepth">https://​www.bioconductor.org/​packages/​megadepth</a>
</li>
<li>Browse source code at <br><a href="https://github.com/LieberInstitute/megadepth/">https://​github.com/​LieberInstitute/​megadepth/​</a>
</li>
<li>Report a bug at <br><a href="https://support.bioconductor.org/t/megadepth">https://​support.bioconductor.org/​t/​megadepth</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/Artistic-2.0">Artistic-2.0</a></li>
</ul>
</div>
<div class="community">
<h2>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
<li><a href="CODE_OF_CONDUCT.html">Code of conduct</a></li>
</ul>
</div>
<div class="citation">
<h2>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html">Citing megadepth</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Leonardo Collado-Torres <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2140-308X" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>David Zhang <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0003-2382-8460" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle: maturing"></a></li>
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/megadepth"><img src="http://www.bioconductor.org/shields/build/release/bioc/megadepth.svg" alt="BioC status"></a></li>
<li><a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/megadepth"><img src="http://www.bioconductor.org/shields/build/devel/bioc/megadepth.svg" alt="BioC dev status"></a></li>
<li><a href="https://codecov.io/gh/LieberInstitute/megadepth?branch=master"><img src="https://codecov.io/gh/LieberInstitute/megadepth/branch/master/graph/badge.svg" alt="Codecov test coverage"></a></li>
<li><a href="https://github.com/LieberInstitute/megadepth/actions"><img src="https://github.com/LieberInstitute/megadepth/workflows/R-CMD-check-bioc/badge.svg" alt="R build status"></a></li>
<li><a href="https://support.bioconductor.org/t/megadepth/"><img src="http://www.bioconductor.org/shields/posts/megadepth.svg" alt="Support site activity, last 6 months: tagged questions/avg. answers per question/avg. comments per question/accepted answers, or 0 if no tagged posts."></a></li>
<li><a href="https://github.com/LieberInstitute/megadepth/issues"><img src="https://img.shields.io/github/issues/LieberInstitute/megadepth" alt="GitHub issues"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Leonardo Collado-Torres, David Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
