<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Compute coverage summarizations across a set of regions — get_coverage • megadepth</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Compute coverage summarizations across a set of regions — get_coverage" />
<meta property="og:description" content="Given an input set of annotation regions, compute coverage summarizations
using Megadepth for a given BigWig file." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">megadepth</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/megadepth.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="../NA/NA/NA/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Compute coverage summarizations across a set of regions</h1>
    <small class="dont-index">Source: <a href='NA/NA/NA/blob/master/R/get_coverage.R'><code>R/get_coverage.R</code></a></small>
    <div class="hidden name"><code>get_coverage.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Given an input set of annotation regions, compute coverage summarizations
using Megadepth for a given BigWig file.</p>
    </div>

    <pre class="usage"><span class='fu'>get_coverage</span>(
  <span class='kw'>bigwig_file</span>,
  op = <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='st'>"sum"</span>, <span class='st'>"mean"</span>, <span class='st'>"max"</span>, <span class='st'>"min"</span>),
  <span class='kw'>annotation</span>,
  prefix = <span class='fu'><a href='https://rdrr.io/r/base/file.path.html'>file.path</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/tempfile.html'>tempdir</a></span>(), <span class='st'>"bw.mean"</span>)
)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>bigwig_file</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> with the path to the input BigWig file.</p></td>
    </tr>
    <tr>
      <th>op</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> specifying the summarization operation to
perform.</p></td>
    </tr>
    <tr>
      <th>annotation</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> path to a BED file.</p></td>
    </tr>
    <tr>
      <th>prefix</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> specifying the output file prefix. This
function creates a file called <code>prefix.annotation.tsv</code> that can be read
again later with <code><a href='read_coverage.html'>read_coverage()</a></code>. By default the file is created in
the <code><a href='https://rdrr.io/r/base/tempfile.html'>tempdir()</a></code> and will be deleted after you close your R session.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>A <a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges-class</a> object with the
coverage summarization across the annotation ranges.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>Note that the chromosome names (seqnames) in the BigWig file and the
annotation file should use the same format. Otherwise, Megadepth will
return 0 counts.</p>
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Coverage functions: 
<code><a href='read_coverage.html'>read_coverage</a>()</code></p></div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>## Install if necessary</span>
<span class='fu'><a href='install_megadepth.html'>install_megadepth</a></span>()
</div><div class='output co'>#&gt; <span class='message'>The latest megadepth version is 1.0.4</span></div><div class='output co'>#&gt; <span class='message'>megadepth has been installed to /github/home/bin</span></div><div class='input'>
<span class='co'>## Locate example BigWig and annotation files</span>
<span class='kw'>example_bw</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"tests"</span>, <span class='st'>"test.bam.all.bw"</span>,
    package = <span class='st'>"megadepth"</span>, mustWork = <span class='fl'>TRUE</span>
)
<span class='kw'>annotation_file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span>(<span class='st'>"tests"</span>, <span class='st'>"testbw2.bed"</span>,
    package = <span class='st'>"megadepth"</span>, mustWork = <span class='fl'>TRUE</span>
)

<span class='co'>## Compute the coverage</span>
<span class='kw'>bw_cov</span> <span class='op'>&lt;-</span> <span class='fu'>get_coverage</span>(<span class='kw'>example_bw</span>, op = <span class='st'>"mean"</span>, annotation = <span class='kw'>annotation_file</span>)
<span class='kw'>bw_cov</span>
</div><div class='output co'>#&gt; GRanges object with 4 ranges and 1 metadata column:
#&gt;         seqnames          ranges strand |       cov
#&gt;            &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
#&gt;   [1]      chr10            0-10      * |     0.000
#&gt;   [2]      chr10 8756697-8756762      * |    15.846
#&gt;   [3]      chr10 4359156-4359188      * |     3.000
#&gt;   [4] GL000219.1   168500-168620      * |     1.258
#&gt;   -------
#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</div><div class='input'>
<span class='co'>## If you want to cast this into a RleList object use the following code:</span>
<span class='co'>## (it's equivalent to rtracklayer::import.bw(as = "RleList"))</span>
<span class='kw'>GenomicRanges</span>::<span class='fu'><a href='https://rdrr.io/pkg/GenomicRanges/man/coverage-methods.html'>coverage</a></span>(<span class='kw'>bw_cov</span>, weight = <span class='st'>"cov"</span>)
</div><div class='output co'>#&gt; RleList of length 2
#&gt; $chr10
#&gt; numeric-Rle of length 8756762 with 4 runs
#&gt;   Lengths: 4359155      33 4397508      66
#&gt;   Values :   0.000   3.000   0.000  15.846
#&gt; 
#&gt; $GL000219.1
#&gt; numeric-Rle of length 168620 with 2 runs
#&gt;   Lengths: 168499    121
#&gt;   Values :  0.000  1.258
#&gt; </div><div class='input'>
<span class='co'>## Checking other tools</span>
<span class='kw'>bed</span> <span class='op'>&lt;-</span> <span class='kw'>rtracklayer</span>::<span class='fu'><a href='https://rdrr.io/pkg/rtracklayer/man/export.html'>import</a></span>(<span class='kw'>annotation_file</span>)
<span class='kw'>bw_cov_rtrack</span> <span class='op'>&lt;-</span> <span class='kw'>rtracklayer</span>::<span class='fu'><a href='https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html'>import.bw</a></span>(<span class='kw'>rtracklayer</span>::<span class='fu'><a href='https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html'>BigWigFile</a></span>(<span class='kw'>example_bw</span>), which = <span class='kw'>bed</span>)
<span class='kw'>bw_cov_rtrack</span>
</div><div class='output co'>#&gt; GRanges object with 11 ranges and 1 metadata column:
#&gt;          seqnames          ranges strand |     score
#&gt;             &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
#&gt;    [1]      chr10            1-10      * |         0
#&gt;    [2]      chr10 4359157-4359159      * |         4
#&gt;    [3]      chr10 4359160-4359185      * |         3
#&gt;    [4]      chr10 4359186-4359188      * |         2
#&gt;    [5]      chr10 8756698-8756699      * |         0
#&gt;    [6]      chr10 8756700-8756714      * |         6
#&gt;    [7]      chr10 8756715-8756761      * |        20
#&gt;    [8]      chr10         8756762      * |         0
#&gt;    [9] GL000219.1   168501-168544      * |         0
#&gt;   [10] GL000219.1   168545-168619      * |         2
#&gt;   [11] GL000219.1          168620      * |         1
#&gt;   -------
#&gt;   seqinfo: 2 sequences from an unspecified genome</div><div class='input'>
<span class='fu'><a href='https://rdrr.io/r/base/names.html'>names</a></span>(<span class='kw'>example_bw</span>) <span class='op'>&lt;-</span> <span class='st'>"example"</span>
<span class='kw'>fullCov</span> <span class='op'>&lt;-</span> <span class='kw'>derfinder</span>::<span class='fu'><a href='https://rdrr.io/pkg/derfinder/man/fullCoverage.html'>fullCoverage</a></span>(<span class='kw'>rtracklayer</span>::<span class='fu'><a href='https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html'>BigWigFileList</a></span>(<span class='kw'>example_bw</span>), chrs = <span class='kw'>GenomeInfoDb</span>::<span class='fu'><a href='https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html'>seqlevels</a></span>(<span class='kw'>bw_cov</span>))
</div><div class='output co'>#&gt; <span class='message'>No methods found in package ‘IRanges’ for request: ‘values&lt;-’ when loading ‘derfinder’</span></div><div class='output co'>#&gt; <span class='message'>No methods found in package ‘IRanges’ for request: ‘values’ when loading ‘derfinder’</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 fullCoverage: processing chromosome chr10</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 loadCoverage: finding chromosome lengths</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 loadCoverage: loading BigWig file /usr/local/lib/R/host-site-library/megadepth/tests/test.bam.all.bw</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 loadCoverage: applying the cutoff to the merged data</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 filterData: originally there were 130694993 rows, now there are 130694993 rows. Meaning that 0 percent was filtered.</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 fullCoverage: processing chromosome GL000219.1</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 loadCoverage: finding chromosome lengths</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 loadCoverage: loading BigWig file /usr/local/lib/R/host-site-library/megadepth/tests/test.bam.all.bw</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 loadCoverage: applying the cutoff to the merged data</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 filterData: originally there were 179198 rows, now there are 179198 rows. Meaning that 0 percent was filtered.</span></div><div class='input'><span class='kw'>regionCov</span> <span class='op'>&lt;-</span> <span class='kw'>derfinder</span>::<span class='fu'><a href='https://rdrr.io/pkg/derfinder/man/getRegionCoverage.html'>getRegionCoverage</a></span>(<span class='kw'>fullCov</span>, regions = <span class='kw'>bed</span>)
</div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 getRegionCoverage: processing chr10</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:17 getRegionCoverage: done processing chr10</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:18 getRegionCoverage: processing GL000219.1</span></div><div class='output co'>#&gt; <span class='message'>2020-09-30 16:18:18 getRegionCoverage: done processing GL000219.1</span></div><div class='input'>
<span class='co'>## We have to round the mean to make them comparable</span>
<span class='kw'>testthat</span>::<span class='fu'><a href='https://testthat.r-lib.org/reference/equality-expectations.html'>expect_equivalent</a></span>(
    <span class='fu'><a href='https://rdrr.io/r/base/Round.html'>round</a></span>(<span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span>(<span class='kw'>regionCov</span>[<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>1</span>, <span class='fl'>3</span><span class='op'>:</span><span class='fl'>4</span>, <span class='fl'>2</span>)], <span class='fu'>function</span>(<span class='kw'>x</span>) <span class='fu'><a href='https://rdrr.io/r/base/mean.html'>mean</a></span>(<span class='kw'>x</span><span class='op'>$</span><span class='kw'>value</span>)), <span class='fl'>3</span>),
    <span class='kw'>bw_cov</span><span class='op'>$</span><span class='kw'>cov</span>,
)

<span class='co'>## Note that there's no need to round here</span>
<span class='kw'>testthat</span>::<span class='fu'><a href='https://testthat.r-lib.org/reference/equality-expectations.html'>expect_equivalent</a></span>(
    <span class='fu'><a href='https://rdrr.io/r/base/lapply.html'>sapply</a></span>(<span class='kw'>regionCov</span>[<span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span>(<span class='fl'>1</span>, <span class='fl'>3</span><span class='op'>:</span><span class='fl'>4</span>, <span class='fl'>2</span>)], <span class='fu'>function</span>(<span class='kw'>x</span>) <span class='fu'><a href='https://rdrr.io/r/base/sum.html'>sum</a></span>(<span class='kw'>x</span><span class='op'>$</span><span class='kw'>value</span>)),
    <span class='fu'>get_coverage</span>(<span class='kw'>example_bw</span>, op = <span class='st'>"sum"</span>, annotation = <span class='kw'>annotation_file</span>)<span class='op'>$</span><span class='kw'>cov</span>,
)
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by David Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


