<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Run Megadepth commands — megadepth_cmd • megadepth</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Run Megadepth commands — megadepth_cmd" />
<meta property="og:description" content="Wrapper functions to run Megadepth commands via
system2('megadepth', ...)." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">megadepth</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../articles/megadepth.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/LieberInstitute/megadepth/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Run Megadepth commands</h1>
    <small class="dont-index">Source: <a href='https://github.com/LieberInstitute/megadepth/blob/master/R/megadepth.R'><code>R/megadepth.R</code></a></small>
    <div class="hidden name"><code>megadepth_cmd.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Wrapper functions to run Megadepth commands via
<code><a href='https://rdrr.io/r/base/system2.html'>system2</a>('megadepth', ...)</code>.</p>
    </div>

    <pre class="usage"><span class='fu'>megadepth_cmd</span><span class='op'>(</span><span class='va'>...</span><span class='op'>)</span>

<span class='fu'>megadepth_shell</span><span class='op'>(</span>input <span class='op'>=</span> <span class='st'>"."</span>, <span class='va'>...</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>...</th>
      <td><p>Arguments to be passed to <code><a href='https://rdrr.io/r/base/system2.html'>system2('megadepth', ...)</a></code>, e.g.
<code>annotation(path)</code> is basically
<code>megadepth_cmd(c('--annotation', path))</code>
(i.e. run the command <code class='command'>megadepth --annotation path</code>).</p></td>
    </tr>
    <tr>
      <th>input</th>
      <td><p>A <code><a href='https://rdrr.io/r/base/character.html'>character(1)</a></code> with the path to the input BAM, BigWig or
text file for Megadepth.</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>See <code><a href='https://rdrr.io/r/base/system2.html'>base::system2()</a></code> for the types of output you can generate.</p>
<p>A <code><a href='https://rdrr.io/r/base/character.html'>character()</a></code> with the capture of the standard output stream
generated by Megadepth.</p>
    <h2 class="hasAnchor" id="functions"><a class="anchor" href="#functions"></a>Functions</h2>

    
<ul>
<li><p><code>megadepth_cmd</code>: Run an arbitrary Megadepth command.</p></li>
<li><p><code>megadepth_shell</code>: Run an arbitrary Megadepth command.</p></li>
</ul>
    <h2 class="hasAnchor" id="references"><a class="anchor" href="#references"></a>References</h2>

    <p><code>megadepth_cmd()</code> is based on blogdown::hugo_cmd() which is available at
<a href='https://github.com/rstudio/blogdown/blob/master/R/hugo.R'>https://github.com/rstudio/blogdown/blob/master/R/hugo.R</a>.</p>
<p><code>megadepth_shell()</code> is based on the shell_ls() example from cmdfun which is
available at <a href='https://snystrom.github.io/cmdfun/index.html'>https://snystrom.github.io/cmdfun/index.html</a>.</p>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'>
<span class='co'>## Install if necessary</span>
<span class='fu'><a href='install_megadepth.html'>install_megadepth</a></span><span class='op'>(</span><span class='op'>)</span>
</div><div class='output co'>#&gt; <span class='message'>The latest megadepth version is 1.0.6</span></div><div class='output co'>#&gt; <span class='message'>megadepth has been installed to /github/home/bin</span></div><div class='input'>
<span class='co'>## Find version</span>
<span class='co'>## megadepth_shell() provides an interface more familiar to R users</span>
<span class='fu'>megadepth_shell</span><span class='op'>(</span>version <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "megadepth 1.0.6"</div><div class='input'><span class='co'>## megadepth_cmd() requires using directly the command line syntax for</span>
<span class='co'>## Megadepth</span>
<span class='fu'>megadepth_cmd</span><span class='op'>(</span><span class='st'>"--version"</span>, stdout <span class='op'>=</span> <span class='cn'>TRUE</span><span class='op'>)</span>
</div><div class='output co'>#&gt; [1] "megadepth 1.0.6"</div><div class='input'>
<span class='co'>## Compare the help files:</span>
<span class='co'># megadepth_shell() captures the standard output and returns a character()</span>
<span class='co'># megadepth_cmd() shows the standard output on the console</span>
<span class='fu'>megadepth_shell</span><span class='op'>(</span><span class='st'>"--help"</span><span class='op'>)</span>
</div><div class='output co'>#&gt;  [1] "megadepth 1.0.6"                                                                                                                                           
#&gt;  [2] ""                                                                                                                                                          
#&gt;  [3] "BAM and BigWig utility."                                                                                                                                   
#&gt;  [4] ""                                                                                                                                                          
#&gt;  [5] "Usage:"                                                                                                                                                    
#&gt;  [6] "  megadepth &lt;bam|bw|-&gt; [options]"                                                                                                                          
#&gt;  [7] ""                                                                                                                                                          
#&gt;  [8] "Options:"                                                                                                                                                  
#&gt;  [9] "  -h --help                Show this screen."                                                                                                              
#&gt; [10] "  --version                Show version."                                                                                                                  
#&gt; [11] "  --threads                # of threads to do: BAM decompression OR compute sums over multiple BigWigs in parallel"                                        
#&gt; [12] "                            if the 2nd is intended then a TXT file listing the paths to the BigWigs to process in parallel"                                
#&gt; [13] "                            should be passed in as the main input file instead of a single BigWig file (EXPERIMENTAL)."                                    
#&gt; [14] "  --prefix                 String to use to prefix all output files."                                                                                      
#&gt; [15] "  --no-auc-stdout          Force all AUC(s) to be written to &lt;prefix&gt;.auc.tsv rather than STDOUT"                                                          
#&gt; [16] "  --no-annotation-stdout   Force summarized annotation regions to be written to &lt;prefix&gt;.annotation.tsv rather than STDOUT"                                
#&gt; [17] "  --no-coverage-stdout     Force covered regions to be written to &lt;prefix&gt;.coverage.tsv rather than STDOUT"                                                
#&gt; [18] "  --keep-order             Output annotation coverage in the order chromosomes appear in the BAM/BigWig file"                                              
#&gt; [19] "                           The default is to output annotation coverage in the order chromosomes appear in the annotation BED file."                       
#&gt; [20] "                           This is only applicable if --annotation is used for either BAM or BigWig input."                                                
#&gt; [21] ""                                                                                                                                                          
#&gt; [22] "BigWig Input:"                                                                                                                                             
#&gt; [23] "Extract regions and their counts from a BigWig outputting BED format if a BigWig file is detected as input (exclusive of the other BAM modes):"            
#&gt; [24] "                                          Extracts all reads from the passed in BigWig and output as BED format."                                          
#&gt; [25] "                                           This will also report the AUC over the annotated regions to STDOUT."                                            
#&gt; [26] "                                           If only the name of the BigWig file is passed in with no other args, it will *only* report total AUC to STDOUT."
#&gt; [27] "  --annotation &lt;bed&gt;                      Only output the regions in this BED applying the argument to --op to them."                                      
#&gt; [28] "  --op &lt;sum[default], mean, min, max&gt;     Statistic to run on the intervals provided by --annotation"                                                      
#&gt; [29] "  --sums-only                             Discard coordinates from output of summarized regions"                                                           
#&gt; [30] "  --bwbuffer &lt;1GB[default]&gt;               Size of buffer for reading BigWig files, critical to use a large value (~1GB) for remote BigWigs."               
#&gt; [31] "                                           Default setting should be fine for most uses, but raise if very slow on a remote BigWig."                       
#&gt; [32] ""                                                                                                                                                          
#&gt; [33] ""                                                                                                                                                          
#&gt; [34] "BAM Input:"                                                                                                                                                
#&gt; [35] "Extract basic junction information from the BAM, including co-occurrence"                                                                                  
#&gt; [36] "If only the name of the BAM file is passed in with no other args, it will *only* report total AUC to STDOUT."                                              
#&gt; [37] "  --junctions          Extract jx coordinates, strand, and anchor length, per read"                                                                        
#&gt; [38] "                       writes to a TSV file &lt;prefix&gt;.jxs.tsv"                                                                                              
#&gt; [39] "  --longreads          Modifies certain buffer sizes to accommodate longer reads such as PB/Oxford."                                                       
#&gt; [40] "  --filter-in          Integer bitmask, any bits of which alignments need to have to be kept (similar to samtools view -f)."                               
#&gt; [41] "  --filter-out         Integer bitmask, any bits of which alignments need to have to be skipped (similar to samtools view -F)."                            
#&gt; [42] ""                                                                                                                                                          
#&gt; [43] "Non-reference summaries:"                                                                                                                                  
#&gt; [44] "  --alts                       Print differing from ref per-base coverages"                                                                                
#&gt; [45] "                               Writes to a CSV file &lt;prefix&gt;.alts.tsv"                                                                                     
#&gt; [46] "  --include-softclip           Print a record to the alts CSV for soft-clipped bases"                                                                      
#&gt; [47] "                               Writes total counts to a separate TSV file &lt;prefix&gt;.softclip.tsv"                                                           
#&gt; [48] "  --only-polya                 If --include-softclip, only print softclips which are mostly A's or T's"                                                    
#&gt; [49] "  --include-n                  Print mismatch records when mismatched read base is N"                                                                      
#&gt; [50] "  --print-qual                 Print quality values for mismatched bases"                                                                                  
#&gt; [51] "  --delta                      Print POS field as +/- delta from previous"                                                                                 
#&gt; [52] "  --require-mdz                Quit with error unless MD:Z field exists everywhere it's"                                                                   
#&gt; [53] "                               expected"                                                                                                                   
#&gt; [54] "  --head                       Print sequence names and lengths in SAM/BAM header"                                                                         
#&gt; [55] ""                                                                                                                                                          
#&gt; [56] "Coverage and quantification:"                                                                                                                              
#&gt; [57] "  --coverage           Print per-base coverage (slow but totally worth it)"                                                                                
#&gt; [58] "  --auc                Print per-base area-under-coverage, will generate it for the genome"                                                                
#&gt; [59] "                       and for the annotation if --annotation is also passed in"                                                                           
#&gt; [60] "                       Defaults to STDOUT, unless other params are passed in as well, then"                                                                
#&gt; [61] "                       if writes to a TSV file &lt;prefix&gt;.auc.tsv"                                                                                           
#&gt; [62] "  --bigwig             Output coverage as BigWig file(s).  Writes to &lt;prefix&gt;.bw"                                                                          
#&gt; [63] "                       (also &lt;prefix&gt;.unique.bw when --min-unique-qual is specified)."                                                                     
#&gt; [64] "                       Requires libBigWig."                                                                                                                
#&gt; [65] "  --annotation &lt;bed&gt;   Path to BED file containing list of regions to sum coverage over"                                                                   
#&gt; [66] "                       (tab-delimited: chrm,start,end)"                                                                                                    
#&gt; [67] "  --op &lt;sum[default], mean&gt;     Statistic to run on the intervals provided by --annotation"                                                                
#&gt; [68] "  --min-unique-qual &lt;int&gt;"                                                                                                                                 
#&gt; [69] "                       Output second bigWig consisting built only from alignments"                                                                         
#&gt; [70] "                       with at least this mapping quality.  --bigwig must be specified."                                                                   
#&gt; [71] "                       Also produces second set of annotation sums based on this coverage"                                                                 
#&gt; [72] "                       if --annotation is enabled"                                                                                                         
#&gt; [73] "  --double-count       Allow overlapping ends of PE read to count twice toward"                                                                            
#&gt; [74] "                       coverage"                                                                                                                           
#&gt; [75] "  --num-bases          Report total sum of bases in alignments processed (that pass filters)"                                                              
#&gt; [76] "  --gzip               Turns on gzipping of coverage output (no effect if --bigwig is passsed),"                                                           
#&gt; [77] "                       this will also enable --no-coverage-stdout."                                                                                        
#&gt; [78] ""                                                                                                                                                          
#&gt; [79] "Other outputs:"                                                                                                                                            
#&gt; [80] "  --read-ends          Print counts of read starts/ends, if --min-unique-qual is set"                                                                      
#&gt; [81] "                       then only the alignments that pass that filter will be counted here"                                                                
#&gt; [82] "                       Writes to 2 TSV files: &lt;prefix&gt;.starts.tsv, &lt;prefix&gt;.ends.tsv"                                                                      
#&gt; [83] "  --frag-dist          Print fragment length distribution across the genome"                                                                               
#&gt; [84] "                       Writes to a TSV file &lt;prefix&gt;.frags.tsv"                                                                                            
#&gt; [85] "  --echo-sam           Print a SAM record for each aligned read"                                                                                           
#&gt; [86] "  --ends               Report end coordinate for each read (useful for debugging)"                                                                         
#&gt; [87] "  --test-polya         Lower Poly-A filter minimums for testing (only useful for debugging/testing)"                                                       
#&gt; [88] ""                                                                                                                                                          
#&gt; [89] ""                                                                                                                                                          </div><div class='input'><span class='fu'>megadepth_cmd</span><span class='op'>(</span><span class='st'>"--help"</span><span class='op'>)</span>
</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Leonardo Collado-Torres, David Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


