<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>megadepth quick start guide • megadepth</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="megadepth quick start guide">
<meta property="og:description" content="megadepth">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">megadepth</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/megadepth.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/LieberInstitute/megadepth/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="megadepth_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>megadepth quick start guide</h1>
                        <h4 class="author">David Zhang</h4>
            <address class="author_afil">
      UCL<br><a class="author_email" href="mailto:#"></a><a href="mailto:david.zhang.12@ucl.ac.uk" class="email">david.zhang.12@ucl.ac.uk</a>
      </address>
                              <h4 class="author">Leonardo Collado-Torres</h4>
            <address class="author_afil">
      Lieber Institute for Brain Development, Johns Hopkins Medical Campus<br><a class="author_email" href="mailto:#"></a><a href="mailto:lcolladotor@gmail.com" class="email">lcolladotor@gmail.com</a>
      </address>
                  
            <h4 class="date">2 October 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/LieberInstitute/megadepth/blob/master/vignettes/megadepth.Rmd"><code>vignettes/megadepth.Rmd</code></a></small>
      <div class="hidden name"><code>megadepth.Rmd</code></div>

    </div>

    
    
<p>The goal of <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> is to provide an R interface to the command line tool <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> for BigWig and BAM related utilities created by <a href="https://twitter.com/chrisnwilks">Christopher Wilks</a> <a id="cite-Wilks_2020"></a>(<a href="https://www.biorxiv.org/content/10.1101/TODO">Wilks, Zhang, Collado-Torres, and Langmead, 2020</a>). This R package enables <strong>fast</strong> processing of BigWig files on downstream packages such as <a href="https://bioconductor.org/packages/dasper">dasper</a> and <a href="https://bioconductor.org/packages/recount3">recount3</a>. The <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> software also provides utilities for processing BAM files and extracting coverage information from them.</p>
<p>Here is an illustration on how fast <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> is compared to other tools for processing local and remote BigWig files.</p>
<div class="figure">
<img src="https://raw.githubusercontent.com/LieberInstitute/megadepth/master/analysis/md_rt_pybw_runtime.png" alt="Timing results. Timing comparison for processing 1,000 genomic regions on a bigWig file that is available on the local disk or on a remote resource. We compared megadepth against rtracklayer and pyBigWig. megadepth is typically faster that these other software solutions for computing the mean coverage across a set of 1,000 input genomic regions. Check &lt;https://github.com/LieberInstitute/megadepth/tree/master/analysis&gt; for more details." width="100%"><p class="caption">
Timing results. Timing comparison for processing 1,000 genomic regions on a bigWig file that is available on the local disk or on a remote resource. We compared megadepth against rtracklayer and pyBigWig. megadepth is typically faster that these other software solutions for computing the mean coverage across a set of 1,000 input genomic regions. Check <a href="https://github.com/LieberInstitute/megadepth/tree/master/analysis" class="uri">https://github.com/LieberInstitute/megadepth/tree/master/analysis</a> for more details.
</p>
</div>
<p>Throughout the documentation we use a capital <code>M</code> to refer to the software by Christopher Wilks and a lower case <code>m</code> to refer to this R/Bioconductor package.</p>
<div id="basics" class="section level1">
<h1 class="hasAnchor">
<a href="#basics" class="anchor"></a>Basics</h1>
<div id="install-megadepth" class="section level2">
<h2 class="hasAnchor">
<a href="#install-megadepth" class="anchor"></a>Install <code>megadepth</code>
</h2>
<p><code>R</code> is an open-source statistical environment which can be easily modified to enhance its functionality via packages. <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> is a <code>R</code> package available via the <a href="http://bioconductor.org">Bioconductor</a> repository for packages. <code>R</code> can be installed on any operating system from <a href="https://cran.r-project.org/">CRAN</a> after which you can install <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> by using the following commands in your <code>R</code> session:</p>
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
      <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
  <span class="op">}</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"megadepth"</span><span class="op">)</span>

<span class="co">## Check that you have a valid Bioconductor installation</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/valid.html">valid</a></span><span class="op">(</span><span class="op">)</span></pre></div>
</div>
<div id="required-knowledge" class="section level2">
<h2 class="hasAnchor">
<a href="#required-knowledge" class="anchor"></a>Required knowledge</h2>
<p><em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> is based on many other packages and in particular in those that have implemented the infrastructure needed for dealing with RNA-seq data and high-throughput sequencing data in general. You might benefit from being familiar with the BigWig file format and the <em><a href="https://bioconductor.org/packages/3.12/rtracklayer">rtracklayer</a></em> for importing those files into R as well as exporting BED files <a id="cite-Lawrence_2009"></a>(<a href="http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract">Lawrence, Gentleman, and Carey, 2009</a>). If you are working with annoation files, <em><a href="https://bioconductor.org/packages/3.12/GenomicFeatures">GenomicFeatures</a></em> and <em><a href="https://bioconductor.org/packages/3.12/GenomicRanges">GenomicRanges</a></em> will also be useful to you.</p>
<p>If you are asking yourself the question “Where do I start using Bioconductor?” you might be interested in <a href="http://lcolladotor.github.io/2014/10/16/startbioc/#.VkOKbq6rRuU">this blog post</a>.</p>
</div>
<div id="asking-for-help" class="section level2">
<h2 class="hasAnchor">
<a href="#asking-for-help" class="anchor"></a>Asking for help</h2>
<p>As package developers, we try to explain clearly how to use our packages and in which order to use the functions. But <code>R</code> and <code>Bioconductor</code> have a steep learning curve so it is critical to learn where to ask for help. The blog post quoted above mentions some but we would like to highlight the <a href="https://support.bioconductor.org/">Bioconductor support site</a> as the main resource for getting help: remember to use the <code>megadepth</code> tag and check <a href="https://support.bioconductor.org/t/megadepth/">the older posts</a>. Other alternatives are available such as creating GitHub issues and tweeting. However, please note that if you want to receive help you should adhere to the <a href="http://www.bioconductor.org/help/support/posting-guide/">posting guidelines</a>. It is particularly critical that you provide a small reproducible example and your session information so package developers can track down the source of the error.</p>
</div>
<div id="citing-megadepth" class="section level2">
<h2 class="hasAnchor">
<a href="#citing-megadepth" class="anchor"></a>Citing <code>megadepth</code>
</h2>
<p>We hope that <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> will be useful for your research. Please use the following information to cite the package and the overall approach. Thank you!</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co">## Citation info</span>
<span class="fu"><a href="https://rdrr.io/r/utils/citation.html">citation</a></span><span class="op">(</span><span class="st">"megadepth"</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Zhang D, Collado-Torres L (2020). _megadepth: BigWig and BAM related</span>
<span class="co">#&gt; utilities_. doi: 10.18129/B9.bioc.megadepth (URL:</span>
<span class="co">#&gt; https://doi.org/10.18129/B9.bioc.megadepth),</span>
<span class="co">#&gt; https://github.com/LieberInstitute/megadepth - R package version</span>
<span class="co">#&gt; 0.99.0, &lt;URL: http://www.bioconductor.org/packages/megadepth&gt;.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Wilks C, Zhang D, Collado-Torres L, Langmead B (2020). "megadepth:</span>
<span class="co">#&gt; BigWig and BAM related utilities." _bioRxiv_. doi: 10.1101/TODO (URL:</span>
<span class="co">#&gt; https://doi.org/10.1101/TODO), &lt;URL:</span>
<span class="co">#&gt; https://www.biorxiv.org/content/10.1101/TODO&gt;.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; To see these entries in BibTeX format, use 'print(&lt;citation&gt;,</span>
<span class="co">#&gt; bibtex=TRUE)', 'toBibtex(.)', or set</span>
<span class="co">#&gt; 'options(citation.bibtex.max=999)'.</span></pre></div>
</div>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick start</h1>
<p>To get started, we need to load the <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> package into our R session. This will load all the required dependencies.</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/LieberInstitute/megadepth">"megadepth"</a></span><span class="op">)</span></pre></div>
<p>Once we have the R package loaded, we need to install the <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> software. We can do so with <code><a href="../reference/install_megadepth.html">install_megadepth()</a></code>, which downloads a binary for your OS (Linux, Windows or macOS) <a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. We can then use with an example BigWig file to compute the coverage at a set of regions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="co">## Install Megadepth</span>
<span class="fu"><a href="../reference/install_megadepth.html">install_megadepth</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; The latest megadepth version is 1.0.4</span>
<span class="co">#&gt; megadepth has been installed to /github/home/bin</span></pre></div>
<p>Next, we might want to use <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> to quantify the coverage at a set of regions of the genome of interest to us. Here we will use two example files that are include in the package for illustration and testing purposes. One of them is a <a href="https://genome.ucsc.edu/goldenPath/help/bigWig.html">bigWig file</a> that contains the base-pair coverage information for a sample of interest and the second one is <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED file</a> which contains the genomic region coordinates of interest. So we first locate them.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="co">## Next, we locate the example BigWig and annotation files</span>
<span class="va">example_bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"tests"</span>, <span class="st">"test.bam.all.bw"</span>,
    package <span class="op">=</span> <span class="st">"megadepth"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>
<span class="va">annotation_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"tests"</span>, <span class="st">"testbw2.bed"</span>,
    package <span class="op">=</span> <span class="st">"megadepth"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co">## Where are they?</span>
<span class="va">example_bw</span>
<span class="co">#&gt; [1] "/__w/_temp/Library/megadepth/tests/test.bam.all.bw"</span>
<span class="va">annotation_file</span>
<span class="co">#&gt; [1] "/__w/_temp/Library/megadepth/tests/testbw2.bed"</span></pre></div>
<p>Once we have located the example files we can proceed to calculating the base-pair coverage for our genomic regions of interest. There are several ways to do this with <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em>, but here we use the user-friendly function <code><a href="../reference/get_coverage.html">get_coverage()</a></code>. This function will perform a given operation <strong>op</strong> on the bigWig file for a given set of regions of interest (<em>annotation</em>). One of those operations is to compute the mean base-pair coverage for each input region. This is what we’ll do with our example bigWig file.</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="co">## We can then use megadepth to compute the coverage</span>
<span class="va">bw_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_coverage.html">get_coverage</a></span><span class="op">(</span>
    <span class="va">example_bw</span>,
    op <span class="op">=</span> <span class="st">"mean"</span>,
    annotation <span class="op">=</span> <span class="va">annotation_file</span>
<span class="op">)</span>
<span class="va">bw_cov</span>
<span class="co">#&gt; GRanges object with 4 ranges and 1 metadata column:</span>
<span class="co">#&gt;         seqnames          ranges strand |     score</span>
<span class="co">#&gt;            &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span>
<span class="co">#&gt;   [1]      chr10            0-10      * |     0.000</span>
<span class="co">#&gt;   [2]      chr10 8756697-8756762      * |    15.846</span>
<span class="co">#&gt;   [3]      chr10 4359156-4359188      * |     3.000</span>
<span class="co">#&gt;   [4] GL000219.1   168500-168620      * |     1.258</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></pre></div>
<p><code><a href="../reference/get_coverage.html">get_coverage()</a></code> returns an object that is familiar to <em><a href="https://bioconductor.org/packages/3.12/GenomicRanges">GenomicRanges</a></em> users, that is, a <code>GRanges</code> object that can be used with other Bioconductor software packages <a id="cite-bioconductor2015"></a>(<a href="https://doi.org/10.1038/nmeth.3252">Huber, Carey, Gentleman, Anders, et al., 2015</a>).</p>
<p>This example is just the tip of the iceberg, as <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> and thus <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> can do a lot of useful processing operations on BAM and bigWig files.</p>
</div>
<div id="users-guide" class="section level1">
<h1 class="hasAnchor">
<a href="#users-guide" class="anchor"></a>Users guide</h1>
<div id="command-interface" class="section level2">
<h2 class="hasAnchor">
<a href="#command-interface" class="anchor"></a>Command interface</h2>
<p><a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> is very powerful and can do a lot of different things. The R/Bioconductor package provides two functions for interfacing with <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a>, <code><a href="../reference/megadepth_cmd.html">megadepth_cmd()</a></code> and <code><a href="../reference/megadepth_cmd.html">megadepth_shell()</a></code>. For the first one, <code><a href="../reference/megadepth_cmd.html">megadepth_cmd()</a></code>, you need to know the actual command syntax you want to use and format it accordingly. If you are more comfortable with R functions, <code><a href="../reference/megadepth_cmd.html">megadepth_shell()</a></code> uses <em><a href="https://CRAN.R-project.org/package=cmdfun">cmdfun</a></em> to power this interface and capture the standard output stream into R.</p>
<p>To make it easier to use, <code>megadepth</code> includes functions that simplify the number of arguments, read in the output files, and converts them into R/Bioconductor friendly objects, such as <code><a href="../reference/get_coverage.html">get_coverage()</a></code> illustrated previously in the quick start section.</p>
<p>We hope that you’ll find <code>megadepth</code> and <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> useful for your work. If you are interested in checking how <strong>fast</strong> <code>megadepth</code> is, check out the <a href="https://github.com/LieberInstitute/megadepth/tree/master/analysis"><strong>speed analysis</strong></a> comparison against other tools. Note that the size of the files used and the number of genomic regions queried will affect the speed comparisons.</p>
<div class="sourceCode" id="cb7"><pre class="downlit">
<span class="co">## R-like interface</span>
<span class="co">## that captures the standard output into R</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/megadepth_cmd.html">megadepth_shell</a></span><span class="op">(</span>help <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; [1] "megadepth 1.0.4"                  ""                                </span>
<span class="co">#&gt; [3] "BAM and BigWig utility."          ""                                </span>
<span class="co">#&gt; [5] "Usage:"                           "  megadepth &lt;bam|bw|-&gt; [options]"</span>

<span class="co">## Command-like interface</span>
<span class="fu"><a href="../reference/megadepth_cmd.html">megadepth_cmd</a></span><span class="op">(</span><span class="st">"--help"</span><span class="op">)</span></pre></div>
<pre><code>#&gt; megadepth 1.0.4
#&gt;  
#&gt;  BAM and BigWig utility.
#&gt;  
#&gt;  Usage:
#&gt;    megadepth &lt;bam|bw|-&gt; [options]
#&gt;  
#&gt;  Options:
#&gt;    -h --help                Show this screen.
#&gt;    --version                Show version.
#&gt;    --threads                # of threads to do: BAM decompression OR compute sums over multiple BigWigs in parallel
#&gt;                              if the 2nd is intended then a TXT file listing the paths to the BigWigs to process in parallel
#&gt;                              should be passed in as the main input file instead of a single BigWig file (EXPERIMENTAL).
#&gt;    --prefix                 String to use to prefix all output files.
#&gt;    --no-annotation-stdout   Force summarized annotation regions to be written to &lt;prefix&gt;.annotation.tsv rather than STDOUT
#&gt;    --no-coverage-stdout     Force covered regions to be written to &lt;prefix&gt;.coverage.tsv rather than STDOUT
#&gt;    --keep-order             Output annotation coverage in the order chromosomes appear in the BAM/BigWig file
#&gt;                             The default is to output annotation coverage in the order chromosomes appear in the annotation BED file.
#&gt;                             This is only applicable if --annotation is used for either BAM or BigWig input.
#&gt;  
#&gt;  BigWig Input:
#&gt;  Extract regions and their counts from a BigWig outputting BED format if a BigWig file is detected as input (exclusive of the other BAM modes):
#&gt;                                            Extracts all reads from the passed in BigWig and output as BED format.
#&gt;                                             This will also report the AUC over the annotated regions to STDOUT.
#&gt;                                             If only the name of the BigWig file is passed in with no other args, it will *only* report total AUC to STDOUT.
#&gt;    --annotation &lt;bed&gt;                      Only output the regions in this BED applying the argument to --op to them.
#&gt;    --op &lt;sum[default], mean, min, max&gt;     Statistic to run on the intervals provided by --annotation
#&gt;    --sums-only                             Discard coordinates from output of summarized regions
#&gt;    --bwbuffer &lt;1GB[default]&gt;               Size of buffer for reading BigWig files, critical to use a large value (~1GB) for remote BigWigs.
#&gt;                                             Default setting should be fine for most uses, but raise if very slow on a remote BigWig.
#&gt;  
#&gt;  
#&gt;  BAM Input:
#&gt;  Extract basic junction information from the BAM, including co-occurrence
#&gt;    --junctions          Extract jx coordinates, strand, and anchor length, per read
#&gt;                         writes to a TSV file &lt;prefix&gt;.jxs.tsv
#&gt;    --longreads          Modifies certain buffer sizes to accommodate longer reads such as PB/Oxford.
#&gt;  
#&gt;  Non-reference summaries:
#&gt;    --alts                       Print differing from ref per-base coverages
#&gt;                                 Writes to a CSV file &lt;prefix&gt;.alts.tsv
#&gt;    --include-softclip           Print a record to the alts CSV for soft-clipped bases
#&gt;                                 Writes total counts to a separate TSV file &lt;prefix&gt;.softclip.tsv
#&gt;    --only-polya                 If --include-softclip, only print softclips which are mostly A's or T's
#&gt;    --include-n                  Print mismatch records when mismatched read base is N
#&gt;    --print-qual                 Print quality values for mismatched bases
#&gt;    --delta                      Print POS field as +/- delta from previous
#&gt;    --require-mdz                Quit with error unless MD:Z field exists everywhere it's
#&gt;                                 expected
#&gt;    --head                       Print sequence names and lengths in SAM/BAM header
#&gt;  
#&gt;  Coverage and quantification:
#&gt;    --coverage           Print per-base coverage (slow but totally worth it)
#&gt;    --auc                Print per-base area-under-coverage, will generate it for the genome
#&gt;                         and for the annotation if --annotation is also passed in
#&gt;                         Writes to a TSV file &lt;prefix&gt;.auc.tsv
#&gt;    --bigwig             Output coverage as BigWig file(s).  Writes to &lt;prefix&gt;.bw
#&gt;                         (also &lt;prefix&gt;.unique.bw when --min-unique-qual is specified).
#&gt;                         Requires libBigWig.
#&gt;    --annotation &lt;bed&gt;   Path to BED file containing list of regions to sum coverage over
#&gt;                         (tab-delimited: chrm,start,end)
#&gt;    --min-unique-qual &lt;int&gt;
#&gt;                         Output second bigWig consisting built only from alignments
#&gt;                         with at least this mapping quality.  --bigwig must be specified.
#&gt;                         Also produces second set of annotation sums based on this coverage
#&gt;                         if --annotation is enabled
#&gt;    --double-count       Allow overlapping ends of PE read to count twice toward
#&gt;                         coverage
#&gt;    --num-bases          Report total sum of bases in alignments processed (that pass filters)
#&gt;  
#&gt;  Other outputs:
#&gt;    --read-ends          Print counts of read starts/ends, if --min-unique-qual is set
#&gt;                         then only the alignments that pass that filter will be counted here
#&gt;                         Writes to 2 TSV files: &lt;prefix&gt;.starts.tsv, &lt;prefix&gt;.ends.tsv
#&gt;    --frag-dist          Print fragment length distribution across the genome
#&gt;                         Writes to a TSV file &lt;prefix&gt;.frags.tsv
#&gt;    --echo-sam           Print a SAM record for each aligned read
#&gt;    --ends               Report end coordinate for each read (useful for debugging)
#&gt;    --test-polya         Lower Poly-A filter minimums for testing (only useful for debugging/testing)
#&gt;  
#&gt; </code></pre>
</div>
<div id="bam-to-bigwig" class="section level2">
<h2 class="hasAnchor">
<a href="#bam-to-bigwig" class="anchor"></a>BAM to bigWig</h2>
<p>One use case of <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> is to convert BAM files to bigWig coverage files. To simplify this process and verify that you are not accidentally overwriting valuable files, <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> provides the function <code><a href="../reference/bam_to_bigwig.html">bam_to_bigwig()</a></code>. To illustrate this functionality, we first locate an example BAM file. We then generate the output bigWig files</p>
<div class="sourceCode" id="cb9"><pre class="downlit">
<span class="co">## Find the example BAM file</span>
<span class="va">example_bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"tests"</span>, <span class="st">"test.bam"</span>,
    package <span class="op">=</span> <span class="st">"megadepth"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co">## Create the BigWig file</span>
<span class="co">## Currently Megadepth does not support this on Windows</span>
<span class="va">example_bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bam_to_bigwig.html">bam_to_bigwig</a></span><span class="op">(</span><span class="va">example_bam</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co">## Path to the output files generated by bam_to_bigwig()</span>
<span class="va">example_bw</span>
<span class="co">#&gt;                            all.bw </span>
<span class="co">#&gt; "/tmp/RtmpT9Djax/test.bam.all.bw"</span></pre></div>
<p>Currently this functionality does not work on Windows. In which case, you can continue the vignette with the following example bigWig file.</p>
<div class="sourceCode" id="cb10"><pre class="downlit">
<span class="co">## On Windows, use the example bigWig file that is already included in</span>
<span class="co">## the R package</span>
<span class="va">example_bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"tests"</span>, <span class="st">"test.bam.all.bw"</span>,
    package <span class="op">=</span> <span class="st">"megadepth"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></pre></div>
</div>
<div id="summarize-coverage-over-regions" class="section level2">
<h2 class="hasAnchor">
<a href="#summarize-coverage-over-regions" class="anchor"></a>Summarize coverage over regions</h2>
<p>Once you have a biWig file, you might want to quantify the mean or total expression across a set of genomic coordinates. bigWig files are typically used by genome browsers, and as part of the <em><a href="https://bioconductor.org/packages/3.12/recount">recount</a></em> and <em><a href="https://bioconductor.org/packages/3.12/recount3">recount3</a></em> projects we have released thousands of them. Before we expand more complex uses cases, you might be interested in <code><a href="../reference/get_coverage.html">get_coverage()</a></code>. This function will use <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> to create a tab-separated value (TSV) file containing the coverage summary information <a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> for a given input file that can be read into R using <code><a href="../reference/read_coverage.html">read_coverage()</a></code> <a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a> as shown below with an example set of genomic regions of interest (<em>annotation</em>).</p>
<div class="sourceCode" id="cb11"><pre class="downlit">
<span class="co">## Next, we locate the example annotation BED file</span>
<span class="va">annotation_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"tests"</span>, <span class="st">"testbw2.bed"</span>,
    package <span class="op">=</span> <span class="st">"megadepth"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co">## Now we can compute the coverage</span>
<span class="va">bw_cov</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_coverage.html">get_coverage</a></span><span class="op">(</span><span class="va">example_bw</span>, op <span class="op">=</span> <span class="st">"mean"</span>, annotation <span class="op">=</span> <span class="va">annotation_file</span><span class="op">)</span>
<span class="va">bw_cov</span>
<span class="co">#&gt; GRanges object with 4 ranges and 1 metadata column:</span>
<span class="co">#&gt;         seqnames          ranges strand |     score</span>
<span class="co">#&gt;            &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;</span>
<span class="co">#&gt;   [1]      chr10            0-10      * |     0.000</span>
<span class="co">#&gt;   [2]      chr10 8756697-8756762      * |    15.846</span>
<span class="co">#&gt;   [3]      chr10 4359156-4359188      * |     3.000</span>
<span class="co">#&gt;   [4] GL000219.1   168500-168620      * |     1.258</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 2 sequences from an unspecified genome; no seqlengths</span></pre></div>
<p>If you are familiar with <em><a href="https://bioconductor.org/packages/3.12/rtracklayer">rtracklayer</a></em>, you’ll notice that the coverage summaries are basically the same to the one that can be generated with <code><a href="https://rdrr.io/pkg/rtracklayer/man/BigWigFile.html">rtracklayer::import.bw()</a></code>, which is what <em><a href="https://bioconductor.org/packages/3.12/derfinder">derfinder</a></em> uses internally.</p>
<div class="sourceCode" id="cb12"><pre class="downlit">
<span class="co">## Checking with derfinder and rtracklayer</span>
<span class="va">bed</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rtracklayer/man/export.html">import</a></span><span class="op">(</span><span class="va">annotation_file</span><span class="op">)</span>

<span class="co">## The file needs a name</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">example_bw</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"example"</span>

<span class="co">## Read in the base-pair coverage data</span>
<span class="va">regionCov</span> <span class="op">&lt;-</span> <span class="fu">derfinder</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/derfinder/man/getRegionCoverage.html">getRegionCoverage</a></span><span class="op">(</span>
    regions <span class="op">=</span> <span class="va">bed</span>,
    files <span class="op">=</span> <span class="va">example_bw</span>,
    verbose <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>
<span class="co">#&gt; Warning: multiple methods tables found for 'rowRanges'</span>
<span class="co">#&gt; No methods found in package 'IRanges' for request: 'values&lt;-' when loading 'derfinder'</span>
<span class="co">#&gt; No methods found in package 'IRanges' for request: 'values' when loading 'derfinder'</span>

<span class="co">## Summarize the base-pair coverage data.</span>
<span class="co">## Note that we have to round the mean to make them comparable.</span>
<span class="fu">testthat</span><span class="fu">::</span><span class="fu"><a href="https://testthat.r-lib.org/reference/equality-expectations.html">expect_equivalent</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">regionCov</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">value</span><span class="op">)</span><span class="op">)</span>, <span class="fl">3</span><span class="op">)</span>,
    <span class="va">bw_cov</span><span class="op">$</span><span class="va">score</span>,
<span class="op">)</span>

<span class="co">## If we compute the sum, there's no need to round</span>
<span class="fu">testthat</span><span class="fu">::</span><span class="fu"><a href="https://testthat.r-lib.org/reference/equality-expectations.html">expect_equivalent</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">sapply</a></span><span class="op">(</span><span class="va">regionCov</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">:</span><span class="fl">4</span>, <span class="fl">2</span><span class="op">)</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">$</span><span class="va">value</span><span class="op">)</span><span class="op">)</span>,
    <span class="fu"><a href="../reference/get_coverage.html">get_coverage</a></span><span class="op">(</span><span class="va">example_bw</span>, op <span class="op">=</span> <span class="st">"sum"</span>, annotation <span class="op">=</span> <span class="va">annotation_file</span><span class="op">)</span><span class="op">$</span><span class="va">score</span>,
<span class="op">)</span></pre></div>
</div>
<div id="bam-to-junctions" class="section level2">
<h2 class="hasAnchor">
<a href="#bam-to-junctions" class="anchor"></a>BAM to junctions</h2>
<p><a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> provides utilities that might be of use for future work or that were developed for building <em><a href="https://bioconductor.org/packages/3.12/recount3">recount3</a></em>. One of these features is the possibility to extract locally co-ocurring junctions from a BAM file as described in the <a href="https://github.com/ChristopherWilks/megadepth#megadepth-pathtobamfile---junctions">Megadepth documentation</a>. This feature works only for junctions for which a read or (read pair) has 2 or more junctions.</p>
<p>To illustrate this functionality, we will use an example BAM file and generate the locally co-occurring junction table with <code><a href="../reference/bam_to_junctions.html">bam_to_junctions()</a></code>. We’ll then read in the data using <code><a href="../reference/read_junction_table.html">read_junction_table()</a></code> <a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit">
<span class="co">## Find the example BAM file</span>
<span class="va">example_bam</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"tests"</span>, <span class="st">"test.bam"</span>,
    package <span class="op">=</span> <span class="st">"megadepth"</span>, mustWork <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span>

<span class="co">## Run bam_to_junctions()</span>
<span class="va">example_jxs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bam_to_junctions.html">bam_to_junctions</a></span><span class="op">(</span><span class="va">example_bam</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co">## Path to the output file generated by bam_to_junctions()</span>
<span class="va">example_jxs</span>
<span class="co">#&gt;                            jxs.tsv </span>
<span class="co">#&gt; "/tmp/RtmpT9Djax/test.bam.jxs.tsv"</span>

<span class="co">## Read the data as a tibble using the format specified at</span>
<span class="co">## https://github.com/ChristopherWilks/megadepth#megadepth-pathtobamfile---junctions</span>
<span class="fu"><a href="../reference/read_junction_table.html">read_junction_table</a></span><span class="op">(</span><span class="va">example_jxs</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 11 x 12</span>
<span class="co">#&gt;    chr      pos strand insert_length cigar coords mate_chr mate_pos mate_strand</span>
<span class="co">#&gt;    &lt;chr&gt;  &lt;int&gt; &lt;chr&gt;          &lt;int&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;       &lt;int&gt; &lt;chr&gt;      </span>
<span class="co">#&gt;  1 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756739 -          </span>
<span class="co">#&gt;  2 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756739 -          </span>
<span class="co">#&gt;  3 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756739 -          </span>
<span class="co">#&gt;  4 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756739 -          </span>
<span class="co">#&gt;  5 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756739 -          </span>
<span class="co">#&gt;  6 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756739 -          </span>
<span class="co">#&gt;  7 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756739 -          </span>
<span class="co">#&gt;  8 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756739 -          </span>
<span class="co">#&gt;  9 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756739 -          </span>
<span class="co">#&gt; 10 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756742 -          </span>
<span class="co">#&gt; 11 chr10 8.76e6 +              23853 47M2… 87567… chr10     8756743 -          </span>
<span class="co">#&gt; # … with 3 more variables: mate_insert_length &lt;int&gt;, mate_cigar &lt;chr&gt;,</span>
<span class="co">#&gt; #   mate_coords &lt;chr&gt;</span></pre></div>
</div>
<div id="teams-involved" class="section level2">
<h2 class="hasAnchor">
<a href="#teams-involved" class="anchor"></a>Teams involved</h2>
<p><em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> was made possible to <a href="https://twitter.com/dyzhang32">David Zhang</a>, the author of <em><a href="https://bioconductor.org/packages/3.12/dasper">dasper</a></em>, and a member of the <a href="https://snca.atica.um.es/">Mina Ryten</a>’s lab at UCL.</p>
<p>The <code>ReCount</code> family involves the following teams:</p>
<ul>
<li>
<a href="http://www.langmead-lab.org/">Ben Langmead’s lab</a> at JHU Computer Science</li>
<li>
<a href="https://www.hansenlab.org/">Kasper Daniel Hansen’s lab</a> at JHBSPH Biostatistics Department</li>
<li>
<a href="http://lcolladotor.github.io/">Leonardo Collado-Torres</a> and <a href="http://aejaffe.com/">Andrew E. Jaffe</a> from <a href="https://www.libd.org/">LIBD</a>
</li>
<li>
<a href="http://nellore.bio/">Abhinav Nellore’s lab</a> at OHSU</li>
<li>
<a href="http://jtleek.com/">Jeff Leek’s lab</a> at JHBSPH Biostatistics Deparment</li>
<li>Data hosted by <a href="https://www.sciserver.org/">SciServer from IDIES at JHU</a>
</li>
</ul>
</div>
<div id="other-related-tools" class="section level2">
<h2 class="hasAnchor">
<a href="#other-related-tools" class="anchor"></a>Other related tools</h2>
<p>The <code>ReCount</code> team has worked on several software solutions and projects that complement each other and enable you to re-use public RNA-seq data. Another Bioconductor package that you might be highly interested in is <em><a href="https://bioconductor.org/packages/3.12/snapcount">snapcount</a></em>, which allows you to use the <a href="http://snaptron.cs.jhu.edu/">Snaptron web services</a>. In particular, <em><a href="https://bioconductor.org/packages/3.12/snapcount">snapcount</a></em> is best for queries over a particular subset of genes or intervals across all or most of the samples in <code>recount2</code>/<code>Snaptron</code>.</p>
<p>We remind you that the <strong>main documentation website</strong> for all the <code>recount3</code>-related projects is available at <a href="https://LieberInstitute/github.io/recount3-docs"><strong>recount.bio</strong></a>. Please check that website for more information about how this R/Bioconductor package and other tools are related to each other.</p>
<p><strong>Thank you!</strong></p>
<p>P.S. An <a href="https://LieberInstitute.github.io/megadepth/">alternative version of this vignette is available</a> that was made using <em><a href="https://CRAN.R-project.org/package=pkgdown">pkgdown</a></em>.</p>
</div>
</div>
<div id="reproducibility" class="section level1">
<h1 class="hasAnchor">
<a href="#reproducibility" class="anchor"></a>Reproducibility</h1>
<p>The <em><a href="https://bioconductor.org/packages/3.12/megadepth">megadepth</a></em> package <a id="cite-Zhang_2020"></a>(<a href="http://www.bioconductor.org/packages/megadepth">Zhang and Collado-Torres, 2020</a>) was made possible thanks to:</p>
<ul>
<li>R <a id="cite-R_Core_Team_2020"></a>(<a href="https://www.R-project.org/">R Core Team, 2020</a>)</li>
<li>
<em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> <a id="cite-Oles_2020"></a>(<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> <a id="cite-Boettiger_2019"></a>(<a href="https://github.com/cboettig/knitcitations">Boettiger, 2019</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> <a id="cite-Xie_2020"></a>(<a href="https://yihui.org/knitr/">Xie, 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> <a id="cite-Allaire_2020"></a>(<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=sessioninfo">sessioninfo</a></em> <a id="cite-Csardi_2018"></a>(<a href="https://github.com/r-lib/sessioninfo#readme">Csárdi, core, Wickham, Chang, et al., 2018</a>)</li>
<li>
<em><a href="https://CRAN.R-project.org/package=testthat">testthat</a></em> <a id="cite-Wickham_2011"></a>(<a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf">Wickham, 2011</a>)</li>
</ul>
<p>This package was developed using <em><a href="https://github.com/lcolladotor/biocthis">biocthis</a></em>.</p>
<p>Code for creating the vignette</p>
<div class="sourceCode" id="cb14"><pre class="downlit">
<span class="co">## Create the vignette</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rstudio/rmarkdown">"rmarkdown"</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/system.time.html">system.time</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rmarkdown/man/render.html">render</a></span><span class="op">(</span><span class="st">"megadepth.Rmd"</span>, <span class="st">"BiocStyle::html_document"</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Extract the R code</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="st"><a href="https://yihui.org/knitr/">"knitr"</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/knitr/man/knit.html">knit</a></span><span class="op">(</span><span class="st">"megadepth.Rmd"</span>, tangle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit">
<span class="co">## Clean up</span>
<span class="fu"><a href="https://rdrr.io/r/base/files.html">file.remove</a></span><span class="op">(</span><span class="st">"megadepth.bib"</span><span class="op">)</span>
<span class="co">#&gt; [1] TRUE</span></pre></div>
<p>Date the vignette was generated.</p>
<pre><code>#&gt; [1] "2020-10-02 19:47:52 UTC"</code></pre>
<p>Wallclock time spent generating the vignette.</p>
<pre><code>#&gt; Time difference of 17.729 secs</code></pre>
<p><code>R</code> session information.</p>
<pre><code>#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  setting  value                       
#&gt;  version  R version 4.0.2 (2020-06-22)
#&gt;  os       Ubuntu 20.04 LTS            
#&gt;  system   x86_64, linux-gnu           
#&gt;  ui       X11                         
#&gt;  language (EN)                        
#&gt;  collate  en_US.UTF-8                 
#&gt;  ctype    en_US.UTF-8                 
#&gt;  tz       Etc/UTC                     
#&gt;  date     2020-10-02                  
#&gt; 
#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────
#&gt;  package              * version    date       lib source                          
#&gt;  AnnotationDbi          1.51.3     2020-07-25 [1] Bioconductor                    
#&gt;  askpass                1.1        2019-01-13 [2] RSPM (R 4.0.0)                  
#&gt;  assertthat             0.2.1      2019-03-21 [2] RSPM (R 4.0.0)                  
#&gt;  backports              1.1.10     2020-09-15 [2] RSPM (R 4.0.2)                  
#&gt;  base64enc              0.1-3      2015-07-28 [2] RSPM (R 4.0.0)                  
#&gt;  bibtex                 0.4.2.3    2020-09-19 [1] RSPM (R 4.0.2)                  
#&gt;  Biobase                2.49.1     2020-09-03 [1] Bioconductor                    
#&gt;  BiocFileCache          1.13.1     2020-08-04 [1] Bioconductor                    
#&gt;  BiocGenerics           0.35.4     2020-06-04 [1] Bioconductor                    
#&gt;  BiocManager            1.30.10    2019-11-16 [2] CRAN (R 4.0.2)                  
#&gt;  BiocParallel           1.23.2     2020-07-06 [1] Bioconductor                    
#&gt;  BiocStyle            * 2.17.1     2020-09-24 [1] Bioconductor                    
#&gt;  biomaRt                2.45.3     2020-09-21 [1] Bioconductor                    
#&gt;  Biostrings             2.57.2     2020-06-09 [1] Bioconductor                    
#&gt;  bit                    4.0.4      2020-08-04 [1] RSPM (R 4.0.2)                  
#&gt;  bit64                  4.0.5      2020-08-30 [1] RSPM (R 4.0.2)                  
#&gt;  bitops                 1.0-6      2013-08-17 [1] RSPM (R 4.0.0)                  
#&gt;  blob                   1.2.1      2020-01-20 [1] RSPM (R 4.0.0)                  
#&gt;  bookdown               0.20       2020-06-23 [1] RSPM (R 4.0.2)                  
#&gt;  BSgenome               1.57.6     2020-09-02 [1] Bioconductor                    
#&gt;  bumphunter             1.31.0     2020-04-28 [1] Bioconductor                    
#&gt;  checkmate              2.0.0      2020-02-06 [1] RSPM (R 4.0.0)                  
#&gt;  cli                    2.0.2      2020-02-28 [2] RSPM (R 4.0.0)                  
#&gt;  cluster                2.1.0      2019-06-19 [3] CRAN (R 4.0.2)                  
#&gt;  cmdfun                 1.0.2      2020-10-02 [1] Github (snystrom/cmdfun@5150060)
#&gt;  codetools              0.2-16     2018-12-24 [3] CRAN (R 4.0.2)                  
#&gt;  colorspace             1.4-1      2019-03-18 [1] RSPM (R 4.0.0)                  
#&gt;  crayon                 1.3.4      2017-09-16 [2] RSPM (R 4.0.0)                  
#&gt;  curl                   4.3        2019-12-02 [2] RSPM (R 4.0.0)                  
#&gt;  data.table             1.13.0     2020-07-24 [1] RSPM (R 4.0.2)                  
#&gt;  DBI                    1.1.0      2019-12-15 [1] RSPM (R 4.0.0)                  
#&gt;  dbplyr                 1.4.4      2020-05-27 [1] RSPM (R 4.0.0)                  
#&gt;  DelayedArray           0.15.13    2020-10-01 [1] Bioconductor                    
#&gt;  derfinder              1.23.1     2020-05-10 [1] Bioconductor                    
#&gt;  derfinderHelper        1.23.1     2020-05-10 [1] Bioconductor                    
#&gt;  desc                   1.2.0      2018-05-01 [2] RSPM (R 4.0.0)                  
#&gt;  digest                 0.6.25     2020-02-23 [2] RSPM (R 4.0.0)                  
#&gt;  doRNG                  1.8.2      2020-01-27 [1] RSPM (R 4.0.0)                  
#&gt;  dplyr                  1.0.2      2020-08-18 [1] RSPM (R 4.0.2)                  
#&gt;  ellipsis               0.3.1      2020-05-15 [2] RSPM (R 4.0.0)                  
#&gt;  evaluate               0.14       2019-05-28 [2] RSPM (R 4.0.0)                  
#&gt;  fansi                  0.4.1      2020-01-08 [2] RSPM (R 4.0.0)                  
#&gt;  foreach                1.5.0      2020-03-30 [1] RSPM (R 4.0.0)                  
#&gt;  foreign                0.8-80     2020-05-24 [3] CRAN (R 4.0.2)                  
#&gt;  Formula                1.2-3      2018-05-03 [1] RSPM (R 4.0.0)                  
#&gt;  fs                     1.5.0      2020-07-31 [2] RSPM (R 4.0.2)                  
#&gt;  generics               0.0.2      2018-11-29 [1] RSPM (R 4.0.0)                  
#&gt;  GenomeInfoDb           1.25.11    2020-09-03 [1] Bioconductor                    
#&gt;  GenomeInfoDbData       1.2.3      2020-10-01 [1] Bioconductor                    
#&gt;  GenomicAlignments      1.25.3     2020-06-09 [1] Bioconductor                    
#&gt;  GenomicFeatures        1.41.3     2020-09-03 [1] Bioconductor                    
#&gt;  GenomicFiles           1.25.0     2020-04-27 [1] Bioconductor                    
#&gt;  GenomicRanges          1.41.6     2020-08-12 [1] Bioconductor                    
#&gt;  ggplot2                3.3.2      2020-06-19 [1] RSPM (R 4.0.1)                  
#&gt;  glue                   1.4.2      2020-08-27 [2] RSPM (R 4.0.2)                  
#&gt;  gridExtra              2.3        2017-09-09 [1] RSPM (R 4.0.0)                  
#&gt;  gtable                 0.3.0      2019-03-25 [1] RSPM (R 4.0.0)                  
#&gt;  highr                  0.8        2019-03-20 [2] RSPM (R 4.0.0)                  
#&gt;  Hmisc                  4.4-1      2020-08-10 [1] RSPM (R 4.0.2)                  
#&gt;  hms                    0.5.3      2020-01-08 [1] RSPM (R 4.0.0)                  
#&gt;  htmlTable              2.1.0      2020-09-16 [1] RSPM (R 4.0.2)                  
#&gt;  htmltools              0.5.0      2020-06-16 [2] RSPM (R 4.0.1)                  
#&gt;  htmlwidgets            1.5.1      2019-10-08 [2] RSPM (R 4.0.0)                  
#&gt;  httr                   1.4.2      2020-07-20 [2] RSPM (R 4.0.2)                  
#&gt;  IRanges                2.23.10    2020-06-13 [1] Bioconductor                    
#&gt;  iterators              1.0.12     2019-07-26 [1] RSPM (R 4.0.0)                  
#&gt;  jpeg                   0.1-8.1    2019-10-24 [1] RSPM (R 4.0.0)                  
#&gt;  jsonlite               1.7.1      2020-09-07 [2] RSPM (R 4.0.2)                  
#&gt;  knitcitations        * 1.0.10     2019-09-15 [1] RSPM (R 4.0.0)                  
#&gt;  knitr                  1.30       2020-09-22 [2] RSPM (R 4.0.2)                  
#&gt;  lattice                0.20-41    2020-04-02 [3] CRAN (R 4.0.2)                  
#&gt;  latticeExtra           0.6-29     2019-12-19 [1] RSPM (R 4.0.0)                  
#&gt;  lifecycle              0.2.0      2020-03-06 [2] RSPM (R 4.0.0)                  
#&gt;  locfit                 1.5-9.4    2020-03-25 [1] RSPM (R 4.0.0)                  
#&gt;  lubridate              1.7.9      2020-06-08 [1] RSPM (R 4.0.2)                  
#&gt;  magrittr               1.5        2014-11-22 [2] RSPM (R 4.0.0)                  
#&gt;  Matrix                 1.2-18     2019-11-27 [3] CRAN (R 4.0.2)                  
#&gt;  MatrixGenerics         1.1.3      2020-09-30 [1] Bioconductor                    
#&gt;  matrixStats            0.57.0     2020-09-25 [1] RSPM (R 4.0.2)                  
#&gt;  megadepth            * 0.99.0     2020-10-02 [1] Bioconductor                    
#&gt;  memoise                1.1.0      2017-04-21 [2] RSPM (R 4.0.0)                  
#&gt;  munsell                0.5.0      2018-06-12 [1] RSPM (R 4.0.0)                  
#&gt;  nnet                   7.3-14     2020-04-26 [3] CRAN (R 4.0.2)                  
#&gt;  openssl                1.4.3      2020-09-18 [2] RSPM (R 4.0.2)                  
#&gt;  pillar                 1.4.6      2020-07-10 [2] RSPM (R 4.0.2)                  
#&gt;  pkgconfig              2.0.3      2019-09-22 [2] RSPM (R 4.0.0)                  
#&gt;  pkgdown                1.6.1.9000 2020-10-01 [1] Github (r-lib/pkgdown@e1e46f2)  
#&gt;  plyr                   1.8.6      2020-03-03 [1] RSPM (R 4.0.2)                  
#&gt;  png                    0.1-7      2013-12-03 [1] RSPM (R 4.0.0)                  
#&gt;  prettyunits            1.1.1      2020-01-24 [2] RSPM (R 4.0.0)                  
#&gt;  progress               1.2.2      2019-05-16 [1] RSPM (R 4.0.0)                  
#&gt;  purrr                  0.3.4      2020-04-17 [2] RSPM (R 4.0.0)                  
#&gt;  qvalue                 2.21.0     2020-04-27 [1] Bioconductor                    
#&gt;  R6                     2.4.1      2019-11-12 [2] RSPM (R 4.0.0)                  
#&gt;  ragg                   0.3.1      2020-07-03 [1] RSPM (R 4.0.2)                  
#&gt;  rappdirs               0.3.1      2016-03-28 [1] RSPM (R 4.0.0)                  
#&gt;  RColorBrewer           1.1-2      2014-12-07 [1] RSPM (R 4.0.0)                  
#&gt;  Rcpp                   1.0.5      2020-07-06 [2] RSPM (R 4.0.2)                  
#&gt;  RCurl                  1.98-1.2   2020-04-18 [1] RSPM (R 4.0.0)                  
#&gt;  readr                  1.3.1      2018-12-21 [1] RSPM (R 4.0.2)                  
#&gt;  RefManageR             1.2.12     2019-04-03 [1] RSPM (R 4.0.0)                  
#&gt;  reshape2               1.4.4      2020-04-09 [1] RSPM (R 4.0.2)                  
#&gt;  rlang                  0.4.7      2020-07-09 [2] RSPM (R 4.0.2)                  
#&gt;  rmarkdown              2.4        2020-09-30 [1] RSPM (R 4.0.2)                  
#&gt;  rngtools               1.5        2020-01-23 [1] RSPM (R 4.0.0)                  
#&gt;  rpart                  4.1-15     2019-04-12 [3] CRAN (R 4.0.2)                  
#&gt;  rprojroot              1.3-2      2018-01-03 [2] RSPM (R 4.0.0)                  
#&gt;  Rsamtools              2.5.3      2020-06-26 [1] Bioconductor                    
#&gt;  RSQLite                2.2.1      2020-09-30 [1] RSPM (R 4.0.2)                  
#&gt;  rstudioapi             0.11       2020-02-07 [2] RSPM (R 4.0.0)                  
#&gt;  rtracklayer            1.49.5     2020-08-07 [1] Bioconductor                    
#&gt;  S4Vectors              0.27.13    2020-09-18 [1] Bioconductor                    
#&gt;  scales                 1.1.1      2020-05-11 [1] RSPM (R 4.0.0)                  
#&gt;  sessioninfo          * 1.1.1      2018-11-05 [2] RSPM (R 4.0.0)                  
#&gt;  stringi                1.5.3      2020-09-09 [2] RSPM (R 4.0.2)                  
#&gt;  stringr                1.4.0      2019-02-10 [2] RSPM (R 4.0.0)                  
#&gt;  SummarizedExperiment   1.19.9     2020-10-01 [1] Bioconductor                    
#&gt;  survival               3.2-7      2020-09-28 [3] RSPM (R 4.0.2)                  
#&gt;  systemfonts            0.3.2      2020-09-29 [1] RSPM (R 4.0.2)                  
#&gt;  testthat               2.3.2      2020-03-02 [2] RSPM (R 4.0.0)                  
#&gt;  tibble                 3.0.3      2020-07-10 [2] RSPM (R 4.0.2)                  
#&gt;  tidyselect             1.1.0      2020-05-11 [1] RSPM (R 4.0.0)                  
#&gt;  utf8                   1.1.4      2018-05-24 [2] RSPM (R 4.0.0)                  
#&gt;  VariantAnnotation      1.35.3     2020-06-26 [1] Bioconductor                    
#&gt;  vctrs                  0.3.4      2020-08-29 [2] RSPM (R 4.0.2)                  
#&gt;  withr                  2.3.0      2020-09-22 [2] RSPM (R 4.0.2)                  
#&gt;  xfun                   0.18       2020-09-29 [2] RSPM (R 4.0.2)                  
#&gt;  XML                    3.99-0.5   2020-07-23 [1] RSPM (R 4.0.2)                  
#&gt;  xml2                   1.3.2      2020-04-23 [2] RSPM (R 4.0.0)                  
#&gt;  XVector                0.29.3     2020-06-25 [1] Bioconductor                    
#&gt;  yaml                   2.2.1      2020-02-01 [2] RSPM (R 4.0.0)                  
#&gt;  zlibbioc               1.35.0     2020-04-27 [1] Bioconductor                    
#&gt; 
#&gt; [1] /__w/_temp/Library
#&gt; [2] /usr/local/lib/R/site-library
#&gt; [3] /usr/local/lib/R/library</code></pre>
</div>
<div id="bibliography" class="section level1">
<h1 class="hasAnchor">
<a href="#bibliography" class="anchor"></a>Bibliography</h1>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.12/BiocStyle">BiocStyle</a></em> (<a href="https://github.com/Bioconductor/BiocStyle">Oleś, Morgan, and Huber, 2020</a>) with <em><a href="https://CRAN.R-project.org/package=knitr">knitr</a></em> (<a href="https://yihui.org/knitr/">Xie, 2020</a>) and <em><a href="https://CRAN.R-project.org/package=rmarkdown">rmarkdown</a></em> (<a href="https://github.com/rstudio/rmarkdown">Allaire, Xie, McPherson, Luraschi, et al., 2020</a>) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=knitcitations">knitcitations</a></em> (<a href="https://github.com/cboettig/knitcitations">Boettiger, 2019</a>).</p>
<p>[1] J. Allaire, Y. Xie, J. McPherson, J. Luraschi, et al. <em>rmarkdown: Dynamic Documents for R</em>. R package version 2.4. 2020. &lt;URL: <a href="https://github.com/rstudio/rmarkdown" class="uri">https://github.com/rstudio/rmarkdown</a>&gt;.</p>
<p>[2] C. Boettiger. <em>knitcitations: Citations for ‘Knitr’ Markdown Files</em>. R package version 1.0.10. 2019. &lt;URL: <a href="https://github.com/cboettig/knitcitations" class="uri">https://github.com/cboettig/knitcitations</a>&gt;.</p>
<p>[3] G. Csárdi, R. core, H. Wickham, W. Chang, et al. <em>sessioninfo: R Session Information</em>. R package version 1.1.1. 2018. &lt;URL: <a href="https://github.com/r-lib/sessioninfo#readme" class="uri">https://github.com/r-lib/sessioninfo#readme</a>&gt;.</p>
<p>[4] W. Huber, V. J. Carey, R. Gentleman, S. Anders, et al. “Orchestrating high-throughput genomic analysis with Bioconductor”. In: <em>Nat Methods</em> (2015). DOI: 10.1038/nmeth.3252.</p>
<p>[5] M. Lawrence, R. Gentleman, and V. Carey. “rtracklayer: an R package for interfacing with genome browsers”. In: <em>Bioinformatics</em> 25 (2009), pp. 1841-1842. DOI: 10.1093/bioinformatics/btp328. &lt;URL: <a href="http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract" class="uri">http://bioinformatics.oxfordjournals.org/content/25/14/1841.abstract</a>&gt;.</p>
<p>[6] A. Oleś, M. Morgan, and W. Huber. <em>BiocStyle: Standard styles for vignettes and other Bioconductor documents</em>. R package version 2.17.1. 2020. &lt;URL: <a href="https://github.com/Bioconductor/BiocStyle" class="uri">https://github.com/Bioconductor/BiocStyle</a>&gt;.</p>
<p>[7] R Core Team. <em>R: A Language and Environment for Statistical Computing</em>. R Foundation for Statistical Computing. Vienna, Austria, 2020. &lt;URL: <a href="https://www.R-project.org/" class="uri">https://www.R-project.org/</a>&gt;.</p>
<p>[8] H. Wickham. “testthat: Get Started with Testing”. In: <em>The R Journal</em> 3 (2011), pp. 5-10. &lt;URL: <a href="https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf" class="uri">https://journal.r-project.org/archive/2011-1/RJournal_2011-1_Wickham.pdf</a>&gt;.</p>
<p>[9] C. Wilks, D. Zhang, L. Collado-Torres, and B. Langmead. “megadepth: BigWig and BAM related utilities”. In: <em>bioRxiv</em> (2020). DOI: 10.1101/TODO. &lt;URL: <a href="https://www.biorxiv.org/content/10.1101/TODO" class="uri">https://www.biorxiv.org/content/10.1101/TODO</a>&gt;.</p>
<p>[10] Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report Generation in R</em>. R package version 1.30. 2020. &lt;URL: <a href="https://yihui.org/knitr/" class="uri">https://yihui.org/knitr/</a>&gt;.</p>
<p>[11] D. Zhang and L. Collado-Torres. <em>megadepth: BigWig and BAM related utilities</em>. <a href="https://github.com/LieberInstitute/megadepth" class="uri">https://github.com/LieberInstitute/megadepth</a> - R package version 0.99.0. 2020. DOI: 10.18129/B9.bioc.megadepth. &lt;URL: <a href="http://www.bioconductor.org/packages/megadepth" class="uri">http://www.bioconductor.org/packages/megadepth</a>&gt;.</p>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>Please check <a href="https://github.com/ChristopherWilks/megadepth">Megadepth</a> for instructions on how to compile the software from source if the binary version doesn’t work for you.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Sum, mean, min or max base-pair coverage for each region.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>If you prefer a <code>tibble</code>, use <code><a href="../reference/read_coverage.html">read_coverage_table()</a></code>.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>The <code>strand</code> columns have been switched from 0s and 1s to <code><a href="https://rdrr.io/r/base/Arithmetic.html">+</a></code> and <code><a href="https://rdrr.io/r/base/Arithmetic.html">-</a></code> for the forard and reverse strands, to match frequently used Bioconductor packages.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Leonardo Collado-Torres, David Zhang.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
